### PR TITLE
mimic: multisite: overwrites in versioning-suspended buckets fail to sync

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/38080

### CODE REVIEW

**Comment by cbodley** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
jenkins test make check (logs are gone again)

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by smithfarm** [comment]
The FTBFS is:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc: In member function 'int RGWRados::repair_olh(RGWObjState*, const RGWBucketInfo&, const rgw_obj&)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:11484:18: error: 'rgw_bl_str' was not declared in this scope
   if (olh.tag == rgw_bl_str(state->olh_tag)) { // mismatch already resolved?
                  ^~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:11484:18: note: suggested alternative: 'rgw_user'
   if (olh.tag == rgw_bl_str(state->olh_tag)) { // mismatch already resolved?
                  ^~~~~~~~~~
                  rgw_user
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:11516:29: error: 'struct rgw_rados_ref' has no member named 'obj'
   r = ref.ioctx.operate(ref.obj.oid, &op);
                             ^~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc: In member function 'int RGWRados::bi_get_olh(const RGWBucketInfo&, const rgw_obj&, rgw_bucket_olh_entry*)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:13079:29: error: 'using bufferlist = class ceph::buffer::list {aka class ceph::buffer::list}' has no member named 'cbegin'; did you mean 'begin'?
   auto iter = bi_entry.data.cbegin();
                             ^~~~~~
                             begin
src/rgw/CMakeFiles/rgw_a.dir/build.make:1434: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/rgw_rados.cc.o' failed
make[3]: *** [src/rgw/CMakeFiles/rgw_a.dir/rgw_rados.cc.o] Error 1
make[3]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:22508: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/all' failed
make[2]: *** [src/rgw/CMakeFiles/rgw_a.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
```

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -11434,6 +11460,61 @@ int RGWRados::bucket_index_read_olh_log(const RGWBucketInfo& bucket_info, RGWObj
...
+                         const rgw_obj& obj)
+{
+  // fetch the current olh entry from the bucket index
+  rgw_bucket_olh_entry olh;
+  int r = bi_get_olh(bucket_info, obj, &olh);
+  if (r < 0) {
+    ldout(cct, 0) << "repair_olh failed to read olh entry for " << obj << dendl;
+    return r;
+  }
+  if (olh.tag == rgw_bl_str(state->olh_tag)) { // mismatch already resolved?
```
Comment (smithfarm) [inline src/rgw/rgw_rados.cc]: @cbodley The FTBFS seems to be emanating from this line - namely, `rgw_bl_str` is not defined. 

To me, that seems to indicate I need to either (a) cherry-pick dcb5e06ca1d or (b) replace `rgw_bl_str(state->olh_tag)` with `state->olh_tag.to_str()`

I'm leaning towards (b)...

Comment (smithfarm) [reply src/rgw/rgw_rados.cc]: . . . taking a stab at cherry-picking dcb5e06 - just for kicks, as an experiment

Comment (smithfarm) [reply src/rgw/rgw_rados.cc]: . . . and I now see that you went with `state->olh_tag.to_str()` in the luminous backport https://github.com/ceph/ceph/pull/26191

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc: In member function 'int RGWRados::repair_olh(RGWObjState*, const RGWBucketInfo&, const rgw_obj&)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:11516:29: error: 'struct rgw_rados_ref' has no member named 'obj'
   r = ref.ioctx.operate(ref.obj.oid, &op);
                             ^~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc: In member function 'int RGWRados::bi_get_olh(const RGWBucketInfo&, const rgw_obj&, rgw_bucket_olh_entry*)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_rados.cc:13079:29: error: 'using bufferlist = class ceph::buffer::list' {aka 'class ceph::buffer::list'} has no member named 'cbegin'; did you mean 'begin'?
   auto iter = bi_entry.data.cbegin();
                             ^~~~~~
                             begin
src/rgw/CMakeFiles/rgw_a.dir/build.make:1434: recipe for target 'src/rgw/CMakeFiles/rgw_a.dir/rgw_rados.cc.o' failed
make[3]: *** [src/rgw/CMakeFiles/rgw_a.dir/rgw_rados.cc.o] Error 1
```

**Comment by smithfarm** [comment]
Note: the backport of dcb5e06ca1d7088b268c7706fe9530e5e99ea579 is https://github.com/smithfarm/ceph/commit/1e97bb4fce - since it failed in the exact same way as shown in the previous comment, I guess it's still a viable option.

**Comment by yuriw** [comment]
[wip-yuri4-testing-2019-07-23-1633-mimic](https://trello.com/c/G8TDc8rz/672-wip-yuri4-testing-2019-07-23-1633-mimic)

**Pull request merged**